File: skills/code-architect/references/code-examples/php/Composite.php
<?php
// Composite — compose objects into trees, treat leaf and container uniformly.
declare(strict_types=1);

interface Node
{
    public function size(): int;
    public function render(int $depth = 0): string;
}

final class FileLeaf implements Node
{
    public function __construct(private string $name, private int $bytes) {}

    public function size(): int
    {
        return $this->bytes;
    }

    public function render(int $depth = 0): string
    {
        return str_repeat('  ', $depth) . "- {$this->name} ({$this->bytes}b)" . PHP_EOL;
    }
}

final class FolderComposite implements Node
{
    /** @var Node[] */
    private array $children = [];

    public function __construct(private string $name) {}

    public function add(Node $child): static
    {
        $this->children[] = $child;
        return $this;
    }

    public function size(): int
    {
        return array_sum(array_map(static fn (Node $n) => $n->size(), $this->children));
    }

    public function render(int $depth = 0): string
    {
        $out = str_repeat('  ', $depth) . "+ {$this->name}/ ({$this->size()}b)" . PHP_EOL;
        foreach ($this->children as $child) {
            $out .= $child->render($depth + 1);
        }
        return $out;
    }
}

// Demo
$root = (new FolderComposite('root'))
    ->add(new FileLeaf('readme.txt', 12))
    ->add(
        (new FolderComposite('src'))
            ->add(new FileLeaf('main.php', 100))
            ->add(new FileLeaf('util.php', 50))
    );

echo $root->render();

File: skills/code-architect/references/code-examples/swift/Composite.swift
// Composite — compose objects into trees, treat leaf and container uniformly.

protocol Node {
    var name: String { get }
    func size() -> Int
}

// Leaf
struct FileNode: Node {
    let name: String
    let bytes: Int
    func size() -> Int { bytes }
}

// Composite
final class FolderNode: Node {
    let name: String
    private var children: [Node] = []

    init(name: String) { self.name = name }

    func add(_ node: Node) {
        children.append(node)
    }

    func size() -> Int {
        children.reduce(0) { $0 + $1.size() }
    }
}

func runDemo() {
    let root = FolderNode(name: "root")
    root.add(FileNode(name: "readme.txt", bytes: 120))

    let sub = FolderNode(name: "src")
    sub.add(FileNode(name: "main.swift", bytes: 800))
    sub.add(FileNode(name: "util.swift", bytes: 450))
    root.add(sub)

    print("total size of \(root.name): \(root.size()) bytes")
    print("size of \(sub.name): \(sub.size()) bytes")
}

runDemo()
